### PR TITLE
Added flapper andflopper to most recent commit of master

### PR DESCRIPTION
**Pull request by nathan111100**

### CODE REVIEW

File: pydss/flapper.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+""""
+//pragma solidity >=0.5.12;
+import "./lib.sol";
+interface VatLike {
+    function move(address,address,uint) external;
+}
+"""
+import time
+class VatLike:
```
Comment (akirillo) [inline pydss/flapper.py:9]: don't worry about interfaces

Comment (akirillo) [reply pydss/flapper.py:9]: and later remove unused interfaces/commented out stuff

File: pydss/flapper.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+   This thing lets you sell some dai in return for gems.
+ - `lot` dai in return for bid
+ - `bid` gems paid
+ - `ttl` single bid lifetime
+ - `beg` minimum bid increase
+ - `end` max auction duration
+*/
+"""
+class Flapper:
+    wards = dict()
```
Comment (akirillo) [inline pydss/flapper.py:36]: don't make these class variables, instantiate in `__init__()`

File: pydss/flapper.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    #messageclass
+    message = 0;
+
+    def rely(self, usr,send):
```
Comment (akirillo) [inline pydss/flapper.py:40]: we don't need auth functions in our sim lol

File: pydss/flapper.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        z = x+y
+        if z > x:
+            return z
+        return -1
+
+    def mul(self, x, y):
+        z = x * y
+        if (y == 0 or (z // y == x)):
+            return z
+        return -1
```
Comment (sehyunc) [inline pydss/flapper.py:93]: we can use built in Python operations instead of defining them

File: pydss/flapper.py (hunk 2 above)
Comment (akirillo) [reply pydss/flapper.py:36]: be sure to do this everywhere

File: pydss/flapper.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+#bid info
+    class Bid :
+        bid = 0;
+        lot = 0 ;
+        guy = 0;
+        tic = 0
+        end = 0
+
+    bids = dict()
+    ONE = 1.00E18;
```
Comment (akirillo) [inline pydss/flapper.py:55]: we'll replace this with constructor calls where necessary

File: pydss/flapper.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    beg = 1.05E18;
```
Comment (akirillo) [inline pydss/flapper.py:56]:
```suggestion
    beg = Wad.from_number(1.05);
```

File: pydss/flapper.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            return z
+        return -1
+
+    def mul(self, x, y):
+        z = x * y
+        if (y == 0 or (z // y == x)):
+            return z
+        return -1
+    def kick(self, lot, bid) :
+        if (self.live != 1):
```
Comment (akirillo) [inline pydss/flapper.py:95]: swap the `ifs` with `require` from `pydss.util`

File: pydss/flapper.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    vat = None;
+    gem = None
+
+    def __init__(self, vat, gem,msg) :
+        self.message = msg
+        self.wards[msg.sender] = 1;
+        self.vat = VatLike(vat);
+        self.gem = GemLike(gem);
+        self.live = 1;
```
[the remark this answers is not in the document]
Comment (akirillo) [reply pydss/flapper.py:81]: add `self.kicks = 0`

File: pydss/flapper.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    #3 hours bidduration[seconds]
+    ttl = 3*60*60;
+    #2 days total auction length  [seconds]
+    tau = 2 *24*60*60
+    kicks = 0;
+    live = 0
+    now = time.time()
+
+    """"
+    // --- Events ---
```
Comment (akirillo) [inline pydss/flapper.py:66]: we're ignoring all events

File: pydss/flapper.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        z = x * y
+        if (y == 0 or (z // y == x)):
+            return z
+        return -1
+    def kick(self, lot, bid) :
+        if (self.live != 1):
+            return "Flapper/not-live"
+        if (self.kicks < -1):
+            return "Flapper/overflow";
+        sid = id(self)
```
Comment (akirillo) [inline pydss/flapper.py:99]:
```suggestion
        self.kicks += 1
        sid = self.kicks
```

File: pydss/flapper.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        sid += self.kicks;
```
Comment (akirillo) [inline pydss/flapper.py:100]:
```suggestion
        self.kicks += 1
        sid = self.kicks
```

File: pydss/flapper.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        if z > x:
+            return z
+        return -1
+
+    def mul(self, x, y):
+        z = x * y
+        if (y == 0 or (z // y == x)):
+            return z
+        return -1
+    def kick(self, lot, bid) :
```
Comment (akirillo) [inline pydss/flapper.py:94]:
```suggestion
    def kick(self, lot, bid, sender, now) :
```

File: pydss/flapper.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (akirillo) [reply pydss/flapper.py:81]: add a `self.ADDRESS = "flapper"`

File: pydss/flapper.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        self.bids[sid].bid = bid;
+        self.bids[sid].lot = lot;
+        self.bids[sid].guy = self.message.sender;
+        self.bids[sid].end = self.add(self.now, self.tau);
+
+        self.vat.move(self.msg.sender, self.address(self), lot);
+        # emit Kick(id, lot, bid);
+    """
+    // --- Admin ---
+    function file(bytes32 what, uint data) external note auth
```
Comment (akirillo) [inline pydss/flapper.py:111]: we do need the admin functions, check `cat.py` for ex.

File: pydss/flapper.py (hunk 10 above)
Comment (akirillo) [inline pydss/flapper.py:99]: let's rename this to `bid_id`

File: pydss/flapper.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        if not (self.bids[sid].tic > self.now or self.bids[sid].tic == 0):
+            return "Flapper/already-finished-tic"
+        if not (self.bids[sid].end > self.now):
+            return "Flapper/already-finished-end"
+
+        if not (lot == self.bids[sid].lot):
+            return "Flapper/lot-not-matching"
+        if not (bid >  self.bids[sid].bid):
+            return "Flapper/bid-not-higher";
+        if not (self.mul(bid, self.ONE) >= self.mul(self.beg, self.bids[sid].bid)):
```
Comment (akirillo) [inline pydss/flapper.py:140]:
```suggestion
        if not (self.mul(bid, self.ONE) >= self.mul(self.beg, self.bids[sid].bid)):
        if not bid >= self.beg * self.bids[bid_id].bid:
```

File: pydss/flapper.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        if not(self.live == 1):
+            return "Flapper/not-live"
+        if not(self.bids[sid].tic != 0 and (self.bids[sid].tic < self.now or self.bids[sid].end < self.now)):
+            return "Flapper/not-finished"
+        self.vat.move(sid, self.bids[sid].guy, self.bids[sid].lot);
+        self.gem.burn(sid, self.bids[sid].bid);
+        del self.bids[sid];
+
+
+    def cage(self, rad):
```
Comment (akirillo) [inline pydss/flapper.py:162]: we're not worrying about emergency shutdown (yet), so can remove `cage` and `yank`

File: pydss/flapper.py [hunk 16]
```diff
@@ -6,6 +6,13 @@
 }
 """
 import time
```
Comment (akirillo) [inline pydss/flapper.py:8]: no need to import this

File: pydss/flapper.py (hunk 1 above)
Comment (akirillo) [reply pydss/flapper.py:16]: bump here, let's remove anything that's unused

File: pydss/flapper.py [hunk 17]
```diff
@@ -33,16 +40,9 @@ def move(self, adr,amt):
...
-
-    def rely(self, usr,send):
-        if self.wards[send] != 1:
-            self.wards[usr] = 1
-    def deny(self, usr,send):
-        if self.wards[send] != 1:
-            self.wards[usr] = 0
+
+
+
```
Comment (akirillo) [inline pydss/flapper.py:45]: could you add documentation for the instance variables in the convention we've set up? here's an example from `Flipper`:

```
class Flipper:
    """
    ADDRESS = str
    beg = Wad
    bids = dict[int: Bid]
    cat = Cat
    ilk_id = str
    kicks = int
    ttl = int
    tau = int
    vat = Vat
    """
```

File: pydss/flapper.py [hunk 18]
```diff
@@ -33,16 +40,9 @@ def move(self, adr,amt):
...
 #bid info
     class Bid :
         bid = 0;
```
Comment (akirillo) [inline pydss/flapper.py:48]: here as well, be sure to remove these static variables and instead instantiate them in an `__init__()` function, and add documentation

File: pydss/flapper.py [hunk 19]
```diff
@@ -70,103 +61,82 @@ class Bid :
       uint256 bid
     );
     """
-    vat = None;
-    gem = None
+
 
     def __init__(self, vat, gem,msg) :
+        self.wards = dict()
```
Comment (akirillo) [inline pydss/flapper.py:67]: no need for `wards`, we're not worried about auth

File: pydss/flapper.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (akirillo) [reply pydss/flapper.py:72]: bump here, still needs `self.ADDRESS` to be set

File: pydss/flapper.py (hunk 12 above)
Comment (akirillo) [reply pydss/flapper.py:86]: bump on this

File: pydss/flapper.py [hunk 20]
```diff
@@ -70,103 +61,82 @@ class Bid :
...
-        self.bids[sid].bid = bid;
-        self.bids[sid].lot = lot;
-        self.bids[sid].guy = self.message.sender;
-        self.bids[sid].end = self.add(self.now, self.tau);
+        require(self.live == 1,"Flapper/not-live")
+
+        require(self.kicks < -1,"Flapper/overflow")
+
+        self.kicks+=1
+        self.bid_id += self.kicks;
```
Comment (akirillo) [inline pydss/flapper.py:92]: `bid_id` shouldn't be a field on `self`, it just pertains to the bid being kicked in the function. this should probably look like:
```suggestion
        bid_id = self.kicks;
```
and all future uses of `self.bid_id` should just be `bid_id`

File: pydss/flapper.py [hunk 21]
```diff
@@ -64,9 +64,9 @@ class Bid :
 
 
     def __init__(self, vat, gem,msg) :
-        self.wards = dict()
-        self.message = msg
-        self.wards[msg.sender] = 1;
+
+        self.msg = msg
```
Comment (akirillo) [inline pydss/flapper.py:68]: we shouldn't be setting a `self.msg`, there's no need for this to be saved to the Flapper instance. all we need is `msg.sender`, and we only need it on line 96 (not in other functions), so you can pass in a `sender` parameter to the function, as per our convention

File: pydss/flapper.py [hunk 22]
```diff
@@ -5,7 +5,7 @@
     function move(address,address,uint) external;
 }
 """
```
Comment (akirillo) [inline pydss/flapper.py:7]: please remove the commented code above this

File: pydss/flapper.py [hunk 23]
```diff
@@ -89,12 +89,12 @@ def kick(self, lot, bid) :
...
+        bid_id = self.kicks;
 
-        self.bids[bid].bid = bid;
-        self.bids[self.bid_id].lot = lot;
-        self.bids[self.bid_id].guy = self.message.sender;
-        self.bids[self.bid_id].end = self.now+self.tau
+        self.bids[bid_id].bid = bid;
+        self.bids[bid_id].lot = lot;
+        self.bids[bid_id].guy = self.msg.sender;
+        self.bids[bid_id].end = self.now+self.tau
```
Comment (akirillo) [inline pydss/flapper.py:97]: `now` won't be an instance variable on `self`, you should pass it in as a parameter to the function

File: pydss/flapper.py [hunk 24]
```diff
@@ -130,8 +130,8 @@ def tend(self, bid_id, lot, bid,sender_id):
         require(self.mul(bid, self.ONE) >= self.mul(self.beg, self.bids[bid_id].bid),"Flapper/insufficient-increase")
 
         if (self.msg.sender != self.bids[bid_id].guy):
-            self.gem.move(self.message.sender, self.bids[bid_id].guy, self.bids[bid_id].bid);
-            self.bids[bid_id].guy = self.message.sender;
+            self.gem.move(self.msg.sender, self.bids[bid_id].guy, self.bids[bid_id].bid);
+            self.bids[bid_id].guy = self.msg.sender;
         self.gem.move(self.msg.sender,sender_id, bid - self.bids[bid_id].bid);
```
Comment (akirillo) [inline pydss/flapper.py:135]: replace this with a `self.gem.transferFrom(sender, self.ADDRESS, bid - self.bids[bid_id].bid)`

File: pydss/flopper.py [hunk 25]
```diff
@@ -56,9 +55,9 @@ class Bid:
         end = 0
 
     def __init__(self, vat, gem, msg):
-        self.wards = dict()
-        self.message = msg
-        self.wards[msg.sender] = 1;
+
+        self.msg = msg
```
Comment (akirillo) [inline pydss/flopper.py:59]: same note as above

File: pydss/flopper.py [hunk 26]
```diff
@@ -92,12 +91,12 @@ def kick(self, gal, lot, bid, now):
     require(self.kicks > -1, "Flapper/overflow")
 
     self.kicks += 1
-    self.bid_id += self.kicks;
+    bid_id += self.kicks;
```
Comment (akirillo) [inline pydss/flopper.py:94]: should just be `bid_id = self.kicks`, not `+=`

File: pydss/flapper.py [hunk 27]
```diff
@@ -93,10 +93,10 @@ def kick(self, lot, bid) :
 
         self.bids[bid_id].bid = bid;
         self.bids[bid_id].lot = lot;
-        self.bids[bid_id].guy = self.msg.sender;
-        self.bids[bid_id].end = self.now+self.tau
+        self.bids[bid_id].guy = sender;
+        self.bids[bid_id].end = now+self.tau
 
-        self.vat.move(self.msg.sender, self.address(self), lot);
+        self.vat.transferfrom(sender, self.ADDRESS, lot);
```
Comment (akirillo) [inline pydss/flapper.py:99]: my bad, could've been clearer - only replace instances of `gem.move` with `gem.transferFrom`, `vat.move` is a method we do have defined on the Vat class and should be called wherever necessary

File: pydss/flapper.py [hunk 28]
```diff
@@ -108,10 +108,10 @@ def file(self,what, data):
             self.tau = data;
         else:
             raise Exception("Flapper/file-unrecognized-param");
-    def deal(self,bid_id):
+    def deal(self,bid_id,sender,bid):
         require(self.live == 1,"Flapper/not-live")
         require(self.bids[bid_id].tic != 0 and (self.bids[bid_id].tic < self.now or self.bids[bid_id].end < self.now),"Flapper/not-finished")
-        self.vat.move(bid_id, self.bids[bid_id].guy, self.bids[bid_id].lot);
+        self.gem.transferFrom(sender, self.ADDRESS, bid - self.bids[bid_id].bid)
```
Comment (akirillo) [inline pydss/flapper.py:114]: ya u can pretty much just undo this change and the change to the function parameters as per the comment above re: `vat.move`

File: pydss/flopper.py [hunk 29]
```diff
@@ -115,7 +115,7 @@ def dent(self, bid_id, lot, bid):
     require(self.lot < self.bids[bid_id].lot,"Flopper/lot-not-lower")
     require(self.beg* lot <= self.bids[bid_id].lot,"Flopper/insufficient-decrease")
     if (self.msg.sender != self.bids[bid_id].guy):
-        self.vat.move(self.bids.msg.sender, self.bids[bid_id].guy, bid);
+        self.vat.transferFrom(sender, self.bids[bid_id].guy, bid);
```
Comment (akirillo) [inline pydss/flopper.py:118]: same here, can revert this change and the change to the function parameters

File: pydss/flopper.py [hunk 30]
```diff
@@ -126,10 +126,10 @@ def dent(self, bid_id, lot, bid):
     self.bids[bid_id].tic = int(self.now)+ self.ttl
 
 
-def deal(self, sid):
+def deal(self, sid,lot):
```
Comment (akirillo) [inline pydss/flopper.py:129]: let's rename `sid` to `bid_id` for consistency, and i believe we don't need the `lot` parameter

File: pydss/flopper.py [hunk 31]
```diff
@@ -126,10 +126,10 @@ def dent(self, bid_id, lot, bid):
...
     require(self.live == 1,"Flapper/not-live")
     require(self.bids[sid].tic != 0 and (self.bids[sid].tic < self.now or self.bids[sid].end < self.now), "Flapper/not-finished")
-    self.vat.move(sid, self.bids[sid].guy, self.bids[sid].lot);
+    self.vat.transferFrom(sid, self.bids[sid].guy, lot);
```
Comment (akirillo) [inline pydss/flopper.py:132]: you can remove this line entirely, in the source code, there's no call to `vat.move`, just the call below to `gem.mint`

File: pydss/flopper.py [hunk 32]
```diff
@@ -126,10 +126,10 @@ def dent(self, bid_id, lot, bid):
     self.bids[bid_id].tic = int(self.now)+ self.ttl
 
 
-def deal(self, sid):
+def deal(self, sid,lot):
     require(self.live == 1,"Flapper/not-live")
     require(self.bids[sid].tic != 0 and (self.bids[sid].tic < self.now or self.bids[sid].end < self.now), "Flapper/not-finished")
```
Comment (akirillo) [inline pydss/flopper.py:131]: looks like you use `self.now` here, be sure to instead accept `now` as a parameter to the function and use it accordingly within

File: pydss/flopper.py [hunk 33]
```diff
@@ -106,7 +106,7 @@ def tick(self, bid_id,now):
     self.bids[bid_id].end = now+ self.tau;
 
 
-def dent(self, bid_id, lot, bid):
+def dent(self, bid_id, lot, bid, sender):
     require( (self.live == 1), "Flopper/not-live")
     require( (self.bids[bid_id].guy != 0),"Flopper/guy-not-set")
     require ( (self.bids[bid_id].tic > self.now or self.bids[id].tic == 0),"Flopper/already-finished-tic")
```
Comment (akirillo) [inline pydss/flopper.py:112]: here as well, using `self.now`
also you've got an extra set of parentheses around the condition

**Comment by akirillo** [comment]
gonna merge this right now, I'll help you get started on the stat trackers cause we already have conventions there
